plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Auction.hs

```diff
@@ -0,0 +1,269 @@
+happen and then deal with the event. The waiting is implemented in
+@waitForChange@ and the event handling is in @handleChange@.
+
+Updates to the user are provided via the log functionality. This is not as bad
+as it sounds because the log records JSON objects so we are at least semi-
+structured. (Basically we have a @Writer [Aeson.Value]@ - in the future we
+should just make it a @Writer a@ for some user-defined @a@ so that we can
+export the "current state" of the app more easily.)
+
+-}
+
+data BuyerChange =
+        AuctionIsOver HighestBid -- ^ The auction has ended with the highest bid
+        | SubmitOwnBid Ada -- ^ We want to submit a new bid
+        | OtherBid HighestBid -- ^ Another buyer submitted a higher bid
+        | NoChange HighestBid -- ^ Nothing has changed
+
+waitForChange :: AuctionParams -> StateMachineClient AuctionState AuctionInput -> HighestBid -> Contract BuyerSchema SM.SMContractError BuyerChange
+waitForChange AuctionParams{apEndTime} client lastHighestBid = do
+    s <- currentSlot
+    let
+        auctionOver = awaitSlot apEndTime >> pure (AuctionIsOver lastHighestBid)
+        submitOwnBid = SubmitOwnBid <$> endpoint @"bid"
+        otherBid = do
+            let address = Scripts.scriptAddress (validatorInstance (SM.scInstance client))
+                targetSlot = succ (succ s) -- FIXME (jm): There is some off-by-one thing going on that requires us to
+                                           -- use succ.succ instead of just a single succ if we want 'addressChangeRequest'
+                                           -- to wait for the next slot to begin.
+                                           -- I don't have the time to look into that atm though :(
+            AddressChangeResponse{acrTxns} <- addressChangeRequest AddressChangeRequest{acreqSlot=targetSlot, acreqAddress = address}
+            case acrTxns of
+                [] -> pure (NoChange lastHighestBid)
+                _  -> currentState client >>= pure . maybe (AuctionIsOver lastHighestBid) OtherBid
+
+    -- see note [Buyer client]
+    auctionOver `select` submitOwnBid `select` otherBid
+
+handleChange :: StateMachineClient AuctionState AuctionInput -> HighestBid -> BuyerChange -> Contract BuyerSchema SM.SMContractError (Either HighestBid ())
+handleChange client lastHighestBid change =
+    let continue = pure . Left
+        stop     = pure (Right ())
+    -- see note [Buyer client]
+    in case change of
+        AuctionIsOver _ -> stop
+        SubmitOwnBid ada -> do
+            self <- Ledger.pubKeyHash <$> ownPubKey
+            r <- SM.runStep client Bid{newBid = ada, newBidder = self}
+            let newHighestBid = HighestBid{highestBid= ada, highestBidder = self}
+            case r of
+                SM.TransitionFailure i -> logError (TransitionFailed i) >> continue lastHighestBid
+                SM.TransitionSuccess _ -> logInfo (BidSubmitted newHighestBid) >> continue newHighestBid
```

Comment:
Doesn't this give us the new state, and hence the new highest bid (which should be ours?). So no need to manually construct it ourselves.